src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -0,0 +1,203 @@
+            Seq.zip items |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
+        | _ -> Seq.zip items [arg] |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
+
+
+/// Returns a new array of the given type and length.
+/// Returns None if the type doesn't have a default value.
+let rec constructNewArray (bt: TypeKind) (length: int): Expr option =
+    match defaultValue bt with
+    | Some x -> ImmutableArray.CreateRange (List.replicate length (wrapExpr x bt)) |> ValueArray |> Some
+    | None -> None
+
+/// Returns the default value for a given type (from Q# documentation)
+and defaultValue (bt: TypeKind): Expr option =
+    match bt with
+    | Int -> IntLiteral 0L |> Some
+    | BigInt -> BigIntLiteral BigInteger.Zero |> Some
+    | Double -> DoubleLiteral 0.0 |> Some
+    | Bool -> BoolLiteral false |> Some
+    | String -> StringLiteral (NonNullable<_>.New "", ImmutableArray.Empty) |> Some
+    | Pauli -> PauliLiteral PauliI |> Some
+    | Result -> ResultLiteral Zero |> Some
+    | Range -> RangeLiteral (wrapExpr (IntLiteral 1L) Int, wrapExpr (IntLiteral 0L) Int) |> Some
+    | ArrayType t -> constructNewArray t.Resolution 0
+    | _ -> None
+
+/// Wraps a QsExpressionType in a basic TypedExpression
+and wrapExpr (expr: Expr) (bt: TypeKind): TypedExpression =
+    let ii = {IsMutable=false; HasLocalQuantumDependency=false}
+    TypedExpression.New (expr, ImmutableDictionary.Empty, ResolvedType.New bt, ii, Null)
+
+
+/// Counts the number of MissingExprs in the given tuple
+let rec countMissingExprs (expr: TypedExpression): int =
+    match expr.Expression with
+    | MissingExpr -> 1
+    | ValueTuple vt -> Seq.map countMissingExprs vt |> Seq.sum
+    | _ -> 0
+    
+/// Replaces any MissingExprs in the given tuple with the first elements of the given list.
+/// Returns the new values of the partial-argument tuple and the argument list.
+let rec fillPartialArg (partialArg: TypedExpression, arg: list<TypedExpression>): TypedExpression * list<TypedExpression> =
+    match partialArg.Expression with
+    | MissingExpr ->
+        match arg with
+        | first::rest -> first, rest
+        | _ -> failwithf "Not enough remaining arguments"
+    | ValueTuple vt ->
+        let newList, newArg =
+            Seq.fold (fun (cpa, ca) v ->
+                let newPa, newCa = fillPartialArg (v, ca)
+                cpa @ [newPa], newCa
+            ) ([], arg) vt
+        let newPa = wrapExpr (ValueTuple (ImmutableArray.CreateRange newList)) partialArg.ResolvedType.Resolution
+        newPa, newArg
+    | _ -> partialArg, arg
+    
+/// Transforms a partially-application call by replacing missing values with the new arguments
+let partialApplyFunction (baseMethod: TypedExpression) (partialArg: TypedExpression) (arg: TypedExpression): Expr =
+    let argsList =
+        if countMissingExprs partialArg = 1 then [arg] else
+        match arg.Expression with
+        | ValueTuple vt ->
+            if countMissingExprs partialArg <> vt.Length
```

Comment:
What about nested argument tuples?

Reply:
These are handled by the fillPartialArg function, the partialApplyFunction just uses the match to deal with the special case of a single argument.